### PR TITLE
Flow: Fix generating arrow functions with param

### PR DESCRIPTION
**Pull request by danharper**
| Q                 | A
| ----------------- | ---
| Bug fix?          | yes
| Breaking change?  | no
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | yes
| Fixed tickets     | #4503, #4211
| License           | MIT
| Doc PR            | -

Fixes:

* `transform-flow-comments` single arrow param would not print type (#4503)

```diff
# const x = (foo: string) => {};
- const x = foo => {};
+ const x = (foo /*: string*/) => {};
```

* `transform-flow-comments` single arrow param would not print optional (#4211)

```diff
# const x = (foo?) => {};
- const x = foo => {};
+ const x = (foo /*:: ?*/) => {};
```

I also found just using `syntax-flow` wouldn't print out types/optionals for arrow functions with a single param. e.g. all of these:

```js
const bar1 = (x: number): string => {};
const bar2 = (x?) => {};
const bar3 = (x?: string) => {};
const bar4 = x => {};
const bar5 = (x): string => {};
const bar6 = (x: number) => {};
const bar7 = <T>(x) => {};
```

(only when they have a single param) output:

```js
const bar1 = x => {};
const bar2 = x => {};
const bar3 = x => {};
const bar4 = x => {};
const bar5 = x => {};
const bar6 = x => {};
const bar7 = x => {};
```

The output now correctly matches the input.

### CODE REVIEW

File: packages/babel-generator/src/generators/methods.js [hunk 1]
```diff
@@ -77,8 +77,14 @@ export function ArrowFunctionExpression(node: Object) {
     this.space();
   }
 
-  if (node.params.length === 1 && t.isIdentifier(node.params[0])) {
-    this.print(node.params[0], node);
+  let param;
+
+  if (
+    node.params.length === 1
+    && t.isIdentifier(param = node.params[0])
```
Comment (hzoo) [inline packages/babel-generator/src/generators/methods.js]: Might be confusing to read to some but makes sense.

**Comment by hzoo** [comment]
Awesome

File: packages/babel-generator/src/generators/methods.js (hunk 1 above)
Comment (danharper) [inline packages/babel-generator/src/generators/methods.js]: ah yeah, idk why I didn't assign it immediately. I've pushed a change for that (and extracted the type checking into a function to be explicit).

**Comment by danez** [comment]
👍  Thanks alot

**Comment by hzoo** [comment]
🎉 great stuff

**Pull request merged**